### ISSUE TITLE
[scheduler cleanup phase 🐼]: Make a Scheduler without configFactory 😱

### ISSUE THREAD

**Issue by misterikkit** [kubernetes/kubernetes#72547]
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.

Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

This is the final goal of [scheduler cleanup phase :panda_face:], and should be done last.

Refactor `scheduler.New` to return a `Scheduler` without using the `factory.configFactory` or `Configurator` types. This will require copying a bunch of logic from `NewConfigFactory` and `configFactory.CreateFromKeys`. All of the other issues in phase :panda_face: are designed to make this possible.

A lot of the logic might remain in `pkg/scheduler/factory`, but it will not involve the `configFactory` type at all. e.g. several methods of `configFactory` have been converted into functions that can be called from `scheduler.New` directly.

Also, we might end up with some duplicated code between `configFactory` and `scheduler.New`. That should be okay since we want to move all users away from `configFactory`.

Also also, this add a lot of logic into `scheduler.go` and we should consider how `New` should be broken into helper functions and how `scheduler.go` might be split into multiple files.

For `PluginFactoryArgs`, it should be sufficient to create one instance directly (all the fields are exported) and use it to get all priorities, predicates, & metadata producers from the factory package. (`getFitPredicateFunctions` and `getPriorityFunctionConfigs` will need to be exported, as well as the metadata producer functions.)

/sig scheduling

/kind cleanup

**Comment by misterikkit** [comment]
/assign @ahmad-diaa  because this relates closely to https://github.com/kubernetes/kubernetes/issues/71860

Let me know if you would prefer someone else to take this one.

**Comment by ahmad-diaa** [comment]
@misterikkit  I'd like to take this one as well.

**Comment by misterikkit** [comment]
@ahmad-diaa If you want to rebase on https://github.com/kubernetes/kubernetes/pull/72614 you should be able to proceed with this fix now. I expect that PR to be merged as-is.

**Comment by alculquicondor** [comment]
This might be easier to perform after #80590

**Comment by ahmad-diaa** [comment]
Just wanted to share my plan regarding the implementation of this issue.
-  `configurator` methods (such as `CreateFromKeys`, `getPluginArgs`, ...) will be changed into exported functions to be used by the `scheduler` and the `configurator` will be one of the parameters. 
- fields will be incrementally removed from the `configurator` and  passed as parameters to the methods. Each of the following would be in a separate PR:
	- framework (already opened this PR #81374)
	- all listers
   - schedulercache
    - podqueue
     - client, volumebinder, 
	- scheduleroptions (hardPodAffinitySymmetricWeight, disablePreemption, percentageOfNodesToScore, bindTimeoutSecond) as well as enableNonPreempting can be removed together.
 - After that the `configurator` type will be empty and it can be removed. The functions could now be moved to the scheduler, unexported and changed into methods for the type `scheduler`.   The `Config` Type could also be removed as well, as it will not be used anymore (`CreateFromKeys` will return a `scheduler` instead).

Would like to know what everyone thinks about this. 

/cc @alculquicondor @draveness

**Comment by alculquicondor** [comment]
cc/ @Huang-Wei

**Comment by alculquicondor** [comment]
cc/ @ahg-g

**Comment by alculquicondor** [comment]
This would be my proposal. Each step can be a single PR or even be composed of some of them.

1. Move all exported types from the `factory` package that are exposed through fields in `scheduler.Scheduler` and place them in `scheduler.go`. Make sure all the fields are actually necessary, for example `GetBinder` is only used internally.
1. Remove all the uses of exported types from the `factory` package that aren't from the `scheduler` package. There might not be any left after the latest PRs.
1. Get rid of the `pkg/scheduler/factory` package:

    1. Move `factory.go` and `plugins.go` (as `factory_algorithms.go`) to the package `pkg/scheduler`. Feel free to propose another name, but `plugins` is confusing given the existence of the framework plugins.
    1. Unexport all functions and types that are only used from `pkg/scheduler`. There seem to be some unused functions already. So we can remove them here. 

1. Get rid of the `factory.Config`, i.e. make the `factory.Configurator` directly create a scheduler.
1. Get rid of the `factory.Configurator`. Migrate all its methods to the `ConfigFactoryArgs`.

There is probably more work left, but I would like to see how it looks at this point. WDYT?

**Comment by ahmad-diaa** [comment]
@alculquicondor sgtm, I did not know we were planning to keep `configFactoryArgs`. I just have a question regarding the following:
>Make sure all the fields are actually necessary, for example GetBinder is only used internally.

Did it mean to make sure whether it is necessary to be kept exported or not? or just use inline functions?

**Comment by alculquicondor** [comment]
> I did not know we were planning to keep configFactoryArgs

Not really. It's just harder to envision that removal at this point. We can reevaluate once the previous steps are done.

> Did it mean to make sure whether it is necessary to be kept exported or not? or just use inline functions?

Yes, exported or not.

**Comment by ahg-g** [comment]
Thanks @ahmad-diaa and @alculquicondor for setting up the plan. Please make sure we don't break CA code when doing that, and untangle any dependencies they may have.

**Comment by alculquicondor** [comment]
Yes, we have opened kubernetes/autoscaler#2402 for that

**Comment by alculquicondor** [comment]
I can think of 2 nexts steps that could happen in any order:

1. Flatten `Configurator` and `ConfigFactoryArgs`: basically make them a single struct by:
  1. exporting the fields in `Configurator` that correspond to `ConfigFactoryArgs`.
  1. Get rid of `NewConfigFactory`. This can be something like `Configurator.setup` that is called before any of the `Configurator.Create*` methods.
1. Make `Configurator.Create` methods return a Scheduler instead of `Config`.

**Comment by ahmad-diaa** [comment]
Do we still need to export the fields even if `factory.go` is now part of the `scheduler` package?
> i. exporting the fields in Configurator that correspond to ConfigFactoryArgs.

**Comment by alculquicondor** [comment]
You are right, we don't. But since we are planning of getting rid of `Configurator` altogether, I think it's fine to export so that we can distinguish input vs calculated fields.

**Comment by ahg-g** [comment]
Ahmad, thanks a lot of all the work you did on cleaning, I think we are done for this release. In the next release we will remove ```algorithm_factory.go``` as part of phase 2 migration of the framework, but ```factory.go``` will stay for a while until we deprecate Policy, we will perhaps rename it to indicate that it represents the policy-to-framework plugin translation layer.

I think we can officially close this.

**Comment by alculquicondor** [comment]
sg
Thanks a lot team
/close
